Add layout test failure analysis script.

bls1999 opened the pull request:
#### 9112a4f172dd3e52a2a2fd386c236243e38fa328
<pre>
Add layout test failure analysis script.
rdar://114117944

Reviewed by Jonathan Bedard.

Adds a layout test failure analysis tool that can be used to easily parse worker logs from a raw layout test `stdio` file (either by loading over a URL or from file). This allows for (among other things) easily outputting a list of tests for a specific worker that can be used when running WebKit tests via `run-webkit-tests`. Complete functionality listed in `--help`.

* Tools/Scripts/layout-test-failure-analysis: Added.

Canonical link: <a href="https://commits.webkit.org/267706@main">https://commits.webkit.org/267706@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fa89ab0209684709a5286115a520aba4fba6adbc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17451 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17776 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18290 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19240 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16331 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17648 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17919 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17977 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/15161 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/20060 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/15211 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15873 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22525 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16216 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16042 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20361 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16630 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/14102 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/15766 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4160 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/20137 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Smackteo reviewed (approved):
I can see the use cases and script is functional.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+
+import re
+import requests
+from urllib.parse import urlparse
```
Should include `webkitpy` first. We don't actually have `requests` `pip installed` everywhere, `webkitpy` will ensure it's on your Python path. This will make the script less stand-alone, but I don't know that we have a use case that would move this script out of the WebKit repository.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+    input_group.add_argument('-u', '-url',
+                        dest='input_url',
+                        help='Specifies the URL from which to retrieve layout test results data.',
+                        type=str)
```
Do we really need two different arguments here? Can't well tell the difference based on the URL having :// in it?

Generally, it's a bit of a "code smell" when we have a mandatory script argument passed via '-<something>'. That's not a hard-and-fast rule, but in this case, it does seem like a single line of Python code could differentiate between a URL and file.

bls1999 replied:
Thanks @briannafan for showing me the wonderful world of positional arguments!

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+                        type=str)
+    parser.add_argument('-l', '--long',
+                        action='store_true',
+                        help='Lengthens the worker output such that it doesn\'t truncate at the requested test. Good for getting a full test list. This has no effect when the test name is omitted.',
```
I'm a bit confused why this is an option. Are we trying to save network bandwidth by not downloading the entire file?

bls1999 replied:
The idea is that we can export the worker in full instead of truncating at a test if one isn't specified.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+# Ensures a directory exists by checking and attempting to create if it doesn't exist (recursively).
+# Returns true if it exists at the end of the attempt, false if not.
+def ensure_directory_exists(dir):
+    if not os.path.isdir(dir):
```
Why do we need this check? Seems like `os.makedirs` with `exist_ok` would imply this, no?

Would also simplify this function because you could imiediately just `return os.path.isdir(dir)`.

bls1999 replied:
This idea is that it attempts to make a directory and returns true/false depending on if it passes/fails. My thinking is this prevents fringe errors with permissions that (if I'm not mistaken) don't get caught by `os.makedirs`.

JonWBedard replied:
That makes sense for what you do on line `101`, but `os.makedirs(dir, exist_ok=True)` will still work even if the directory already exists. What I'm saying is that you want this, I think:

```
def ensure_directory_exists(dir):
    os.makedirs(dir, exist_ok=True)
    return os.path.isdir(dir)
```

bls1999 replied:
Good point!

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+# Takes a string and exports to specified file path.
+# Prompts the user for a new file path if there's an IO error.
+def export_file(path, data, no_warn = False):
+    path
```
This statement doesn't do anything.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+            print(path)
+            f = open(path, "w")
+            f.write(data)
+            f.close()
```
Should use `with open(path, "w")` here instead. Will simplify your code and gurantees cleanup in the event that the write fails.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+        except IsADirectoryError:
+            warning(f'The specified path "{path}" is a directory. Try entering a new file path.')
+        except OSError:
+            warning(f'The specified location "{path}" is read-only. Try entering a new file path.')
```
This seems like a pretty rare case, and not the only reason we might throw an `OSError`. We should either log the `OSError` itself or just let our caller deal with the exception.

bls1999 replied:
If you're a typoist such as myself, you might type something like `/Volumes/MyDriv/path/to/file.txt` instead of `MyDrive`. That's where I imagine this would come in. The other way is if you're trying to write to a directory that you don't have permission to enter.

I was considering that there might be other `OSError` instances, but couldn't really think of any in this specific case that would break execution such that we should be exiting. Do you have any?

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+            warning(f'The specified path "{path}" is a directory. Try entering a new file path.')
+        except OSError:
+            warning(f'The specified location "{path}" is read-only. Try entering a new file path.')
+        except ValueError as e:
```
No reason to use `as e` if we aren't going to use the `e`. Also, in what circumstance would a `ValueError` be raised? In Python, `ValueError` is usually quite analagous to a compile error, it's often better to let such exception crash the script. In any case, certainly should not surpress this error.

bls1999 replied:
Good catch on `as e`, I think that was leftover from debugging.

The reason this is here is to handle a manual `ValueError` invocation earlier in the function. This would be thrown in the case that the user doesn't want to overwrite the file. There shouldn't be any other circumstance in which a `ValueError` is raised.

I originally programmed this such that there was a `finally` block, but python is weird with `finally` and `return`, so it would've executed the `finally` block even after `return` was called after a successful write operation. To fix this, I opted for the thrown `ValueError` approach.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+        except ValueError as e:
+            pass
+        except Exception as e:
+            warning(f'Unable to write file: {e.args[1]}. Try entering a new file path.')
```
Again, no reason to use `as e` if we aren't going to use the `e`. We usually avoid catching generic `Exception`s, unless we're working in code which is managing other code (like our multiprocessing code, for example). The error message kind of demonstrates why. We really have no way of knowing what the raised exception is about, so we can't really suggest recovery steps.

bls1999 replied:
`e` is used in the string, `f'Unable to write file: {e.args[1]}. Try entering a new file path.'`.

The idea here is that, if the input file has already been downloaded and parsed, and all that needs to be done is write the output to a file, we don't want to crash and make the user start all over again. After the exception, the user is prompted to enter a new file path.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+    else: # all workers
+        num_workers = len(data.keys())
+        if not no_warn:
+            choice = input(f'{num_workers} worker file(s) are about to be output. Is it okay to bypass individual file overwrite confirmation messages? (Y/[N]): ').lower()
```
Since you need to import `webkitpy` anyways, might consider `webkitcorepy`'s `Terminal.choose` here.

bls1999 replied:
Nifty!

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+
+# Loads the stdio input for analysis from a build URL.
+# Returns a str with the contents of the stdio log.
+def load_from_url(url):
```
Specifically, you're loading from a buildbot URL. Which is probably important to specify in this function name.

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+        data = requests.get(url)
+    except Exception as e:
+        error('Unable to load data: ' + e.args[0] + '. Check your spelling and try again.')
+        graceful_exit(os.EX_DATAERR)
```
This is a good example of the right way to catch a general exception.

Worth noting that `sys.exit` itself actually raises an exception. This code isn't broken by that, but it's a bit of a surprising fact that can bite defensive error handling like this which attempts to manage general exceptions.

bls1999 replied:
Thanks for the info on that! Does `sys.exit` raising an exception break this, and/or does it require any changes?

JonWBedard replied:
`sys.exit` raising an exception won't break this, just wanted to call that fact out since I can see refactors which would be bit by that.

bls1999 replied:
Gotcha, I'll add a comment to that effect. Thanks for the tip!

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,291 @@
+
+
+def main():
+    logging.basicConfig(level=logging.INFO, format="%(message)s")
```
Should look into `webkitcorepy`'s `LoggingGroup`, which will let users control your scripts logging level with `-v` and `-q`.

bls1999 replied:
Nifty!

JonWBedard commented on Tools/Scripts/layout-test-failure-analysis:
```diff
@@ -0,0 +1,286 @@
+
+    graceful_exit()
+
+main()
```
You want to guard this in case someone wants to come back and write tests:
```
if '__main__' == __name__:
    main()
```

bls1999 committed:
Add layout test failure analysis script.
rdar://114117944

Reviewed by Jonathan Bedard.

Adds a layout test failure analysis tool that can be used to easily parse worker logs from a raw layout test `stdio` file (either by loading over a URL or from file). This allows for (among other things) easily outputting a list of tests for a specific worker that can be used when running WebKit tests via `run-webkit-tests`. Complete functionality listed in `--help`.

* Tools/Scripts/layout-test-failure-analysis: Added.

Canonical link: https://commits.webkit.org/267706@main
